File: assignments/11-bitwise/src/Easy/Q18.go
package main

import "fmt"

func subsetXORSum(nums []int) int {
	subSet := make([][]int, 1)

	for i := 0; i < len(nums); i++ {
		len := len(subSet)
		for j := 0; j < len; j++ {
			xx := make([]int, 0)
			xx = append(xx, subSet[j]...)
			xx = append(xx, nums[i])
			subSet = append(subSet, xx)
		}
	}

	//fmt.Println(subSet)
	sum := 0
	for _, arr := range subSet {
		xor := 0
		for i := 0; i < len(arr); i++ {
			xor ^= arr[i]
		}
		sum += xor
	}

	return sum
}

func main() {
	fmt.Println(subsetXORSum([]int{5,1,6}))
}

File: assignments/11-bitwise/src/Easy/Q18.java
package Easy;

import java.util.ArrayList;

public class Q18 {
    public int subsetXORSum(int[] nums) {
        ArrayList<ArrayList<Integer>> subSets = new ArrayList<>();
        subSets.add(new ArrayList<>());

        for (int i = 0 ; i<nums.length; i++) {
            int len = subSets.size();

            for (int j = 0; j < len; j++) {
                ArrayList<Integer> x = new ArrayList<>(subSets.get(j));
                x.add(nums[i]);
                subSets.add(x);
            }
        }
        int sum = 0;
        for (ArrayList<Integer> arr : subSets) {
            int xor = 0;
            for (Integer nn : arr) {
                xor ^= nn;
            }
            sum += xor;
        }
        return sum;
    }

    public static void main(String[] args) {
        System.out.println(new Q18().subsetXORSum(new int[] {5,1,6}));
    }

}
